pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionSnapshotCache.java

```diff
@@ -32,32 +34,114 @@
 
-    public synchronized void addNewSnapshot(ReplicatedSubscriptionsSnapshot snapshot) {
+    /**
+     * Adds a new replicated subscription snapshot to the cache with time-based eviction policies.
+     * This method handles 5 distinct cases for cache management:
+     * <ol>
+     *   <li><b>Empty Cache</b>: Directly inserts the snapshot if cache is empty.</li>
+     *   <li><b>Position Precedes First Entry</b>: Clears entire cache and inserts new snapshot.</li>
+     *   <li><b>Position in Middle</b>: Trims later entries and inserts new snapshot.</li>
+     *   <li><b>Cache Not Full</b>: Inserts snapshot if cache has available space.</li>
+     *   <li><b>Cache Full</b>: Evicts median-aged entry before insertion (time-window optimized).</li>
+     * </ol>
+     *
+     * <p>Time-based eviction considers two conditions:
+     * <ul>
+     *   <li>Minimum frequency interval ({@code snapshotFrequencyMillis})</li>
+     *   <li>Dynamic time window per slot ({@code timeWindowPerSlot})</li>
+     * </ul>
+     *
+     * @param snapshot    The replicated subscription snapshot to add (non-null)
+     * @param publishTime The ReplicatedSubscriptionsSnapshot marker message (entry),
+     *                    the publishTime originates from the same broker, the time is reliable and clock skew is
+     *                    not a problem.
+     * @throws NullPointerException If snapshot argument is null
+     * @see ReplicatedSubscriptionsSnapshot
+     * @see #findPositionByIndex(int)  For median position calculation in eviction case
+     */
+    public synchronized void addNewSnapshot(ReplicatedSubscriptionsSnapshot snapshot, long publishTime) {
+        SnapshotEntry snapshotEntry = new SnapshotEntry(snapshot, publishTime);
         MarkersMessageIdData msgId = snapshot.getLocalMessageId();
         Position position = PositionFactory.create(msgId.getLedgerId(), msgId.getEntryId());
 
         if (log.isDebugEnabled()) {
             log.debug("[{}] Added new replicated-subscription snapshot at {} -- {}", subscription, position,
                     snapshot.getSnapshotId());
         }
+        // Case 1: cache if empty
+        if (snapshots.lastEntry() == null) {
+            snapshots.put(position, snapshotEntry);
+            return;
+        }
 
-        snapshots.put(position, snapshot);
+        // The time difference between the previous position and the earliest cache entry
+        final long timeSinceFirstSnapshot = publishTime - snapshots.firstEntry().getValue().timestamp();
+        // The time difference between the previous position and the lately cache entry
+        final long timeSinceLastSnapshot = publishTime - snapshots.lastEntry().getValue().timestamp();
+        // The time window length of each time slot, used for dynamic adjustment in the snapshot cache.
+        // The larger the time slot, the slower the update.
+        final long timeWindowPerSlot = timeSinceFirstSnapshot / snapshotFrequencyMillis / maxSnapshotToCache;
```

Comment:
this is a bit hard to grasp, what the "time slot" concept is here. 
Let's say, timeSinceFirstSnapshot is 25000 ms, snapshotFrequencyMillis is 1000ms and maxSnapshotToCache is 10, it would result in `2`. What's the point of this?
With low values, this would be close to 0, I guess. This is also why I think this is just unnecessary complexity.

Reply:
What if timeSinceFirstSnapshot is 25 minutes?

Reply:
> What if timeSinceFirstSnapshot is 25 minutes? The goal is that if timeSinceFirstSnapshot becomes longer, the update frequency should be lower.

This doesn't seem to be a realistic case. I disagree that the update frequency should become lower. That's exactly my point that if there's a long delay, it will be enforced going forward. I think it's easier to make progress in this PR by removing this rule and adding it later if there's a specific reason to do so.